include/swf.h

```diff
@@ -0,0 +1,254 @@
+	uint16_t width;
+	uint16_t height;
+	uint8_t *data;
+};
+
+// We only support simple rectangle shapes with a fill style.
+struct swf_tag_define_shape {
+	struct swf_tag t;
+	uint16_t shape_id;
+	struct swf_rect bounds;
+	struct swf_fill_style fill_style;
+};
+
+struct swf_tag_define_sound {
+	struct swf_tag t;
+	uint16_t sound_id;
+	uint8_t format;
+	struct swf_sound_spec spec;
+	uint32_t sample_count;
+	uint8_t *data;
+	uint32_t data_len;
+};
+
+struct swf_tag_define_sprite {
+	struct swf_tag t;
+	uint16_t sprite_id;
+	uint16_t frame_count;
+	struct swf_tag *tags;
+};
+
+struct swf_tag_do_action {
+	struct swf_tag t;
+	struct swf_action *actions;
+};
+
+enum place_object_flags {
+	PLACE_FLAG_RESERVED_BITS       = 0xffff - ((1 << 11) - 1),
+	PLACE_FLAG_HAS_CACHE_AS_BITMAP = 1 << 10,
+	PLACE_FLAG_HAS_BLEND_MODE      = 1 << 9,
+	PLACE_FLAG_HAS_FILTER_LIST     = 1 << 8,
+	PLACE_FLAG_HAS_CLIP_ACTIONS    = 1 << 7,
+	PLACE_FLAG_HAS_CLIP_DEPTH      = 1 << 6,
+	PLACE_FLAG_HAS_NAME            = 1 << 5,
+	PLACE_FLAG_HAS_RATIO           = 1 << 4,
+	PLACE_FLAG_HAS_COLOR_TRANSFORM = 1 << 3,
+	PLACE_FLAG_HAS_MATRIX          = 1 << 2,
+	PLACE_FLAG_HAS_CHARACTER       = 1 << 1,
+	PLACE_FLAG_MOVE                = 1 << 0,
+};
+
+struct swf_tag_place_object {
+	struct swf_tag t;  // TAG_PLACE_OBJECT2 or TAG_PLACE_OBJECT3
+	uint16_t flags;
+	uint16_t depth;
+	uint16_t character_id;
+	struct swf_matrix matrix;
+	struct swf_cxform_with_alpha color_transform;
+	uint16_t ratio;
+	char *name;
+	uint8_t blend_mode;
+	uint8_t bitmap_cache;
+};
+
+struct swf_tag_remove_object2 {
+	struct swf_tag t;
+	uint16_t depth;
+};
+
+struct swf_tag_set_background_color {
+	struct swf_tag t;
+	rgb_t background_color;
+};
+
+struct swf_tag_sound_stream_head {
+	struct swf_tag t;
+	struct swf_sound_spec playback_spec;
+	uint8_t compression;
+	struct swf_sound_spec stream_spec;
+	uint16_t sample_count;
+	int16_t latency_seek;
+};
+
+enum start_sound_flags {
+	SOUND_INFO_RESERVED_BITS    = 0xff - ((1 << 6) - 1),
+	SOUND_INFO_SYNC_STOP        = 1 << 5,
+	SOUND_INFO_SYNC_NO_MULTIPLE = 1 << 4,
+	SOUND_INFO_HAS_ENVELOPE     = 1 << 3,
+	SOUND_INFO_HAS_LOOPS        = 1 << 2,
+	SOUND_INFO_HAS_OUT_POINT    = 1 << 1,
+	SOUND_INFO_HAS_IN_POINT     = 1 << 0,
+};
+
+struct swf_tag_start_sound {
+	struct swf_tag t;
+	uint16_t sound_id;
+	uint8_t flags;
+	uint32_t in_point;
+	uint32_t out_point;
+	uint16_t loop_count;
+};
+
+// Actions
+
+enum swf_action_type {
+	// Actions without payload.
+	ACTION_END = 0x00,
+	ACTION_PLAY = 0x06,
+	ACTION_STOP = 0x07,
+	ACTION_POP = 0x17,
+	ACTION_CALL_FUNCTION = 0x3d,
+
+	// Actions with payloads.
+	ACTION_GOTO_FRAME = 0x81,
+	ACTION_CONSTANT_POOL = 0x88,
+	ACTION_PUSH = 0x96,
+};
+
+struct swf_action {
+	uint8_t code;
+	struct swf_action *next;
+};
+
+struct swf_action_constant_pool {
+	struct swf_action a;
+	uint16_t count;
+	char *constant_pool[];
+};
+
+struct swf_action_goto_frame {
+	struct swf_action a;
+	uint16_t frame;
+};
+
+enum swf_actipn_push_type {
```

Comment:
Looks like there's a typo here

Reply:
Good catch, fixed and rebased.

Also added a commit that updates libsys4.